### ISSUE TITLE
[HOLD for payment 2024-07-17] [$250] [Held requests] Incorrectly showing the held confirmation modal for one expense report

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#41613]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.70-1
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @JmillsExpensify 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1714590485812579

## Action Performed:
Prerequisite: Create a Collect workspace as user A, invite user B and configure to submit to user A
1. As user B create an expense
2. As user A, hold user B’s expense
3. Approve the report

## Expected Result:
Hold confirmation modal should not be displayed in the case of only one expense
## Actual Result:
Confirmation modal displayed 
## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [X] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
![CleanShot 2024-05-01 at 13 04 44@2x](https://github.com/Expensify/App/assets/38435837/c41c5ede-c09f-4e56-bcac-4af66b2577d4)
![Screen Shot 2024-05-03 at 3 32 12 PM (2)](https://github.com/Expensify/App/assets/38435837/fab31d57-6173-4dde-a808-f036600f9f73)

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0147c167b8b6e9da09</li>
        <li>Upwork Job ID: 1787747404592803840</li>
        <li>Last Price Increase: 2024-05-21</li>
<li>Automatic offers: </li>
<ul>
<li>bernhardoj | Contributor | 102500989</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @puneetlath</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The approve confirmation shows even all the request is on hold.

### What is the root cause of that problem?
I don't know how to hold a single transaction, but I can reproduce it when hold all transactions in an expense report. I guess the expectation is that the confirmation modal is not needed when all transactions are on hold.

When we press the approve button, it checks whether the report has at least one on hold transaction.
https://github.com/Expensify/App/blob/2f5173c9c0f041967acfbdf9581ce52327623fa6/src/components/MoneyReportHeader.tsx#L138-L145

So, when we have all transactions on hold, the condition is still true and the confirmation modal is shown instead of immediately approving it.

### What changes do you think we should make in order to solve the problem?

We can update the condition to
```
if (ReportUtils.hasHeldExpenses(moneyRequestReport.reportID) && !ReportUtils.hasOnlyHeldExpenses(moneyRequestReport.reportID)) {
```

So, it will only show the confirmation modal if the report contains an on hold transaction and a non hold transaction.

We can do the same for pay here
https://github.com/Expensify/App/blob/2f5173c9c0f041967acfbdf9581ce52327623fa6/src/components/MoneyReportHeader.tsx#L125-L136

**Comment by laurenreidexpensify** [comment]
@mananjadhav bump for review thanks

**Comment by mananjadhav** [comment]
> @JmillsExpensify do we need to consider the expected results here? Should it have different modal copy to confirm you're okay with taking the expense off-hold and approving it?

I was waiting for the confirmation on this comment. @trjExpensify @JmillsExpensify any updates on this one?

**Comment by trjExpensify** [comment]
By "hold pattern" do you mean the banner copy yeah? If so, yeah, that should be there though you can't hold a one expense report right now until that gets fixed.

Even still, I think it would be consistent that in all cases of either one or 20 expenses on the report, if something is held on the report and you go to take an action to approve/pay you would get a confirmation modal. Kinda' even more so in the one-expense report case, as it's an unusual action to take to approve or pay an expense that is held. 

Not super passionate, but I think we're giving people a lot of credit with the below sentiment. #nobodyreads, haha. 

>If you see something is on hold, including via the comments, we shouldn't ask you yet again if you're sure.

**Comment by laurenreidexpensify** [comment]
@Expensify/design any thoughts?

**Comment by shawnborton** [comment]
Hmm I don't feel too strongly, happy to leave the decision up to the other designers + Jason/Tom :)

**Comment by dannymcclain** [comment]
>Even still, I think it would be consistent that in all cases of either one or 20 expenses on the report, if something is held on the report and you go to take an action to approve/pay you would get a confirmation modal. Kinda' even more so in the one-expense report case, as it's an unusual action to take to approve or pay an expense that is held.

I think I'm with Tom on this one. And I tend to agree with this:

>Not super passionate, but I think we're giving people a lot of credit with the below sentiment. #nobodyreads, haha.
>
>>If you see something is on hold, including via the comments, we shouldn't ask you yet again if you're sure.

So I think I'm in favor of leaving the confirmation modal in all cases.

**Comment by dubielzyk-expensify** [comment]
> So I think I'm in favor of leaving the confirmation modal in all cases.

Just catching up but I'm aligned with showing the modal here.

**Comment by laurenreidexpensify** [comment]
Okay so it sounds like the path recommended by @trjExpensify @dannymcclain and @dubielzyk-expensify is to to show the modal. which means this issue can be closed and :donothing: - correct?

**Comment by JmillsExpensify** [comment]
We can't do nothing, as the modal is incorrect. It's asking if you want to approve all or some, but there is only one option. So I think those who chimed in above are proposing that we create a special confirmation modal for this case.

**Comment by JmillsExpensify** [comment]
More generally though, I'm cool with saying that we'll always show the confirmation modal in all cases for now – so one-expense report, in the report preview, and in the report thread.

**Comment by dannymcclain** [comment]
>We can't do nothing, as the modal is incorrect. It's asking if you want to approve all or some, but there is only one option.

Very true. Would it be easiest to update the copy for this situation? Could it say _something like_:

---

**Confirm approval**
All expenses on the report are on hold. Do you want to approve it?

**[ Approve $500.00 ]**

---

☝️ Needs copy finessing, but hopefully you get the idea.

**Comment by trjExpensify** [comment]
I think that's great personally, and yeah, that's what I meant originally by using a modal with copy that makes sense for this scenario. 👍

**Comment by dubielzyk-expensify** [comment]
Yeah I like that copy update. I think the predictability of the same behaviour is mostly what I care about

**Comment by laurenreidexpensify** [comment]
Sounds good! @bernhardoj do you have a suggestion for what that solution would look like?

**Comment by bernhardoj** [comment]
I have updated my [proposal](https://github.com/Expensify/App/issues/41613#issuecomment-2094058984) to conditionally use a different message when all expenses are on hold.

I think the only thing left is the exact copy and its translation. (I'm assuming [this](https://github.com/Expensify/App/issues/41613#issuecomment-2121266984) is not the copy that we are going to use, right?)

**Comment by laurenreidexpensify** [comment]
@shmaxey can you take a look at this and confirm if the copy @dannymcclain suggested here - https://github.com/Expensify/App/issues/41613#issuecomment-2121266984 - is good? thanks

**Comment by shmaxey** [comment]
I think we should be clear that approving the report means all of the expenses will no longer be on hold - 

>All expenses on this report are currently on hold. Do you want to approve the report, removing the hold status on all expenses?

**Comment by laurenreidexpensify** [comment]
how about - 

All expenses on this report are on hold. Do you want to unhold them and approve?

[Approve $500.00]

**Comment by dannymcclain** [comment]
@laurenreidexpensify That sounds good to me!

**Comment by trjExpensify** [comment]
Love it. It's a bit shorter and conveys the point about unholding if you approve. 👍

**Comment by laurenreidexpensify** [comment]
@mananjadhav let's assign @bernhardoj 👍

**Comment by mananjadhav** [comment]
> @mananjadhav let's assign @bernhardoj 👍

Yes agreed.

🎀 👀 🎀 C+ Reviewed.

**Comment by jasperhuangg** [comment]
@bernhardoj asked for you in Slack about the question–I'll respond once I know more.

**Comment by jasperhuangg** [comment]
Hey back with a question–so "All expenses on this report are on hold. Do you want to unhold them and approve?" is what we want for when every expense on the report is held, but I'm a bit confused on what we want the wording be for when there is a mix of both held and unheld expenses on the report.

It seems currently we have the following: "Approve what's not on hold, or approve the entire report", but IMO this wording is a bit confusing, because we're not given an option to select which choice.

**Comment by trjExpensify** [comment]
>Hey back with a question–so "All expenses on this report are on hold. Do you want to unhold them and approve?" is what we want for when every expense on the report is held,

Yep, that's correct. 

>but I'm a bit confused on what we want the wording be for when there is a mix of both held and unheld expenses on the report.

No change to the existing modal copy in production.

**Comment by laurenreidexpensify** [comment]
@puneetlath adding you here for parental leave sub. no action required this BZ this week, PR is still in review, but adding you for payment in coming weeks when the fix makes it to prod. Thanks

**Comment by jasperhuangg** [comment]
Updated the PR with Spanish translations, they also just have some Typescript errors to look into and we should be very close to merging this.

**Comment by bernhardoj** [comment]
I'll request in ND once payment is due.

**Comment by bernhardoj** [comment]
The payment should be due [tomorrow](https://github.com/Expensify/App/issues/41613#issuecomment-2220201600) (07-17)

**Comment by puneetlath** [comment]
Yep. @mananjadhav friendly reminder about the checklist.

**Comment by mananjadhav** [comment]
@puneetlath This changed with a few iterations and I don't have an exact offending PR. But we should add regression tests for this. @puneetlath @jasperhuangg  What do you think of the following from the PR test steps?

**Prerequisite: approval enabled**
 
 1. As user A, submit 2 expenses to B's workspace
 2. As user A, hold one of the expenses
 3. As user B, open the policy expense report
 4. As User B, press approve button
 5. As user B, verify the confirmation modal message shows as `Approve what's not on hold, or approve the entire report.` (unchanged)
 6. As user A, hold all expenses
 7. As user B, press approve button
 8. As user B, verify the confirmation modal message shows as `These expenses are on hold. Do you want to approve anyway?`
 9. As user A, delete one of the expenses
 10. As user B, press approve button
 11. As user B, verify the confirmation modal message shows as `This expense is on hold. Do you want to approve anyway?`
 
 **Prerequisite: approval disabled**
 
 1. Repeat steps 1-3 above
 2. As user B, press pay button
 3. As user B, verify the confirmation modal message shows as `Pay what's not on hold, or pay the entire report.`
 4. As user A, hold all expenses
 5. As user B, press pay button
 6. As user B, verify the confirmation modal message shows as `These expenses are on hold. Do you want to pay anyway?`
 7. As user A, delete one of the expenses
 8. As user B, press pay button
 9. As user B, verify the confirmation modal message shows as `This expense is on hold. Do you want to pay anyway?`

**Comment by puneetlath** [comment]
These are great, thanks!

**Comment by bernhardoj** [comment]
Requested in ND.

**Comment by puneetlath** [comment]
Payment summary: 
- C - @bernhardoj - $250 - to be paid in NewDot
- C+ - @mananjadhav - $250 - to be paid in NewDot

Regression test issue: https://github.com/Expensify/Expensify/issues/413388

**Comment by JmillsExpensify** [comment]
$250 approved for @bernhardoj

**Comment by JmillsExpensify** [comment]
$250 approved for @mananjadhav
